app/models/annotation_merger.rb

```diff
@@ -0,0 +1,104 @@
+class AnnotationMerger
+  def initialize(annotations)
+    # If annotation["text"] is nil, set it to an empty string.
+    # If the text ends with a period, add a space to clarify sentence boundaries when merging annotations.
+    # This prevents loss of readability or meaning.
+    # Setting nil to an empty string also avoids errors in subsequent processing.
+    # Ensure consistent denotations and relations format for reliable merging.
+    @annotations = annotations.map do |annotation|
+      text = annotation["text"] || ""
+      text += " " if text.end_with?(".")
+      denotations = annotation["denotations"] || []
+      relations = annotation["relations"] || []
+      annotation.merge("text" => text, "denotations" => denotations, "relations" => relations)
+    end
+  end
+
+  def merged
+    result = { "text" => merged_text }
+    result["denotations"] = merged_denotations if merged_denotations.any?
+    result["relations"] = merged_relations if merged_relations.any?
+    result
+  end
+
+  private
+
+  # Pre-calculate information for each chunk (length and offset)
+  def chunks_info
+    @chunks_info ||= @annotations.each_with_object([]).with_index do |(annotation, chunks_info), index|
+      text = annotation["text"]
+      # Padding check is not necessary because preprocessing is already done
+      offset = if chunks_info.last
+        chunks_info.last[:offset] + chunks_info.last[:length]
+      else
+        0
+      end
+      chunks_info << {
+        text:,
+        length: text.length,
+        offset:
+      }
+    end
```

Comment:
```suggestion
      text = annotation["text"]
      # Padding check is not necessary because preprocessing is already done
      offset = if chunks_info.last
        chunks_info.last[:offset] + chunks_info.last[:length]
      else
        0
      end

      chunks_info << {
        text:,
        length: text.length,
        offset:
      }
    end
```

このように、コードのあとに chunks_info を戻り値として返す場合、その直前に空行を入れておくことをおすすめします。

Ruby では return を省略するケースが一般的ですが、その分「どの値が返されているのか」が読み取りにくくなることがあります。
戻り値の直前に空行を挟むことで、意図的に返している値であることが視覚的に明確になり、読み手の理解を助けます。

特に戻り値の処理が複数行にわたる場合に有効です。

Reply:
https://github.com/jdkim/textae_campus/pull/19/commits/f17cd925f8151d702ea6d0874dc200291caef829 で修正しました。